### PR TITLE
Scripts started via CLI should also be able to write into the process table

### PR DESCRIPTION
**Pull request by Bezkup**
Hi! 
## Reference 
With this PR, I'm providing a feature that allows the script launched via CLI to be saved in Process table.
Fixes #8775
## Instructions for Reviewers

This feature is disabled by default. To activate it you need to go to dspace.cfg and set `process.save-enable = true`, and most important, you need to set the EPerson ID to associate the owner process in `process.eperson`.

## Description 
- I modified the steps of CommandLineDSpaceRunnableHandler to manage the creation of process on database (like RestDspaceRunnableHandler). 
- I created a new constructor for CommandLineDSpaceRunnableHandler to save script name and the parameters list

PS: this is my first PR, feel free to suggest me how to improve my PR or its commit :)

- [x] My PR is small in size (e.g. less than 1,000 lines of code, not including comments & integration tests). Exceptions may be made if previously agreed upon.
- [x] My PR passes Checkstyle validation based on the [Code Style Guide](https://wiki.lyrasis.org/display/DSPACE/Code+Style+Guide).
- [x] My PR includes Javadoc for all new (or modified) public methods and classes. It also includes Javadoc for large or complex private methods.
- [x] My PR passes all tests and includes new/updated Unit or Integration Tests based on the [Code Testing Guide](https://wiki.lyrasis.org/display/DSPACE/Code+Testing+Guide).
- [ ] If my PR includes new libraries/dependencies (in any pom.xml), I've made sure their licenses align with the [DSpace BSD License](https://github.com/DSpace/DSpace/blob/main/LICENSE) based on the [Licensing of Contributions](https://wiki.lyrasis.org/display/DSPACE/Code+Contribution+Guidelines#CodeContributionGuidelines-LicensingofContributions) documentation.
- [ ] If my PR modifies REST API endpoints, I've opened a separate [REST Contract](https://github.com/DSpace/RestContract/blob/main/README.md) PR related to this change.
- [x] If my PR includes new configurations, I've provided basic technical documentation in the PR itself.
- [x] If my PR fixes an issue ticket, I've [linked them together](https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue).

### CODE REVIEW

File: dspace/config/dspace.cfg [hunk 1]
```diff
@@ -1585,6 +1585,17 @@ solr-database-resync.time-until-reindex = 600000
...
 solr-database-resync.cron = 0 15 2 * * ?
 
+#----------------------------------------------------------#
+#----------PROCESS CONFIGURATION------------#
+#----------------------------------------------------------#
+# Configuration to enable saving script via CLI in DB
+# default = false
+#process.save-enable = false
+# !!! if process.save is enabled, make sure that this field has a correct EPerson's email
+#process.eperson=
```
Comment (tdonohue) [inline dspace/config/dspace.cfg:1595]: Just a quick comment.  I understand why this is necessary, as it appears an [EPerson is *required* to create a Process](https://github.com/DSpace/DSpace/blob/main/dspace-api/src/main/java/org/dspace/scripts/Process.java#L54-L56).  However, I still find this a bit clunky.  I'm worried that it's misleading that CLI processes could appear under a EPerson who didn't run the process.

But, I admit, I don't have a better "quick solution". The only way to make this work without a config would be to either:
1. Change Process to allow for a null EPerson
2. Or, require all Scripts to specify an EPerson (via a flag) when run from CLI.

Neither of those seem like quick solutions though.  But, I wanted to mention them here in case it prompts ideas from other developers.

Comment (Bezkup) [reply dspace/config/dspace.cfg:1595]: The first one seems more efficient than "force" users to set a flag everytime, especially those that have a lot of crontabs to manage Dspace process. I'll see how this null Eperson field impact in Dspace Process, if you agree.

Comment (tdonohue) [reply dspace/config/dspace.cfg:1595]: I'm OK with investigating the effort to allow for a `null` EPerson in a Process.  But, I'll admit, I have not dug into this concept deeply at all, so I don't know whether it will have negative side effects.

Comment (Bezkup) [reply dspace/config/dspace.cfg:1575]: Hi! I'm back. After a deep look, I've seen that setting `null` EPerson doesn't seem to have side effects. 
The only place where it might causes a NPE is [here](https://github.com/DSpace/DSpace/blob/9cbfad4df837d17fa99147d99ea285a0f2f47097/dspace-server-webapp/src/main/java/org/dspace/app/rest/converter/ProcessConverter.java#L42l) because of `process.getEPerson().getID()`.

Comment (tdonohue) [reply dspace/config/dspace.cfg:1575]: @Bezkup : In that case, if you want to update this PR to use a `null` EPerson, and fix any code where an NPE might occur, that seems reasonable.  

I would also recommend "documenting" that Eperson may be null.  There's two logical way to do that.  One is to add tests (or ITs) which prove that a null EPerson is "valid" (i.e. doesn't throw errors)... that might be the best approach anyhow in proving this enhancement "works".   The other would be to update JavaDocs to document this (if it's too difficult to find a place to add a test where Eperson is set to null).

Overall, this sounds like the approach we'd want for this PR if you have time to clean this up in the near future.  

As an FYI, we are in the review process for 8.0 PRs. So if you can find time to update this in the next week or two, it'd allow us to consider this for inclusion. If you don't have time soon, we can do our best to move this forward when you have time...but whether it gets into 8.0 may depend on whether we still have time to safely merge it.  Thanks!

Comment (Bezkup) [reply dspace/config/dspace.cfg]: @tdonohue Nice! I think I'm done (waiting for your feedbacks) :) 
I added a test case for create a `Process` without `EPerson` in `ProcessIT` class, hope it's the correct class to put it.

**Review by tdonohue — CHANGES_REQUESTED** [review]
@Bezkup : I gave this a test today, and I cannot get it to work with the User Interface.  Whenever I attempt to save a script started from CLI, it appears to break the Processes page of the User Interface (the backend error is a NullPointerException)

Here's what I tried:
1. Installed this PR for the backend, using the latest frontend from the `main` branch
2. Enabled the feature via `process.save-enabled = true`
3. Run a process via commandline (e.g. `./dspace index-discovery -b`)
4. Login to the UI as an Admin, and go to the Processes page.  The page will not load.  In the backend logs I see this error:
```
2024-02-05 22:35:15,335 ERROR 023c6910-8f69-45fd-bfa1-5e2b69e0d5c2 1237fac5-dd61-47cb-ac81-f5591b85e80b org.dspace.app.rest.exception.DSpaceApiExceptionControllerAdvice @ An exception has occurred (status:500)
 java.lang.NullPointerException: null
   at org.dspace.app.rest.converter.ProcessConverter.convert(ProcessConverter.java:42) ~[dspace-server-webapp-8.0-SNAPSHOT.jar:8.0-SNAPSHOT]
   at org.dspace.app.rest.converter.ProcessConverter.convert(ProcessConverter.java:24) ~[dspace-server-webapp-8.0-SNAPSHOT.jar:8.0-SNAPSHOT]
   at org.dspace.app.rest.converter.ConverterService.toRest(ConverterService.java:112) ~[dspace-server-webapp-8.0-SNAPSHOT.jar:8.0-SNAPSHOT]
   at org.dspace.app.rest.converter.ConverterService.toRestPage(ConverterService.java:235) ~[dspace-server-webapp-8.0-SNAPSHOT.jar:8.0-SNAPSHOT]
   at org.dspace.app.rest.converter.ConverterService$$FastClassBySpringCGLIB$$41367eae.invoke(<generated>) ~[dspace-server-webapp-8.0-SNAPSHOT.jar:8.0-SNAPSHOT]
```

It looks like the bug is here: https://github.com/DSpace/DSpace/blob/main/dspace-server-webapp/src/main/java/org/dspace/app/rest/converter/ProcessConverter.java#L42  The code assumes that EPerson will not be null.

We may need to verify that all areas of the code that call `process.getEPerson()` are able to handle that being null.

I have a few other minor notes inline below.  The code looks good overall.  I can retest it once it is working again.

File: dspace-api/src/main/java/org/dspace/scripts/ProcessServiceImpl.java [hunk 2]
```diff
@@ -92,10 +92,17 @@ public Process create(Context context, EPerson ePerson, String scriptName,
...
-                                          scriptName + " and parameters " + parameters));
+
+        if (ePerson != null) {
+            log.info(LogHelper.getHeader(context, "process_create",
+                "Process has been created for eperson with email " + ePerson.getEmail()
+                    + " with ID " + createdProcess.getID() + " and scriptName " +
+                    scriptName + " and parameters " + parameters));
+        } else {
+            log.info(LogHelper.getHeader(context, "process_create",
+                "Process has been created for anonymous user with ID " + createdProcess.getID()
```
[the remark this answers is not in the document]
Comment (Bezkup) [reply dspace-api/src/main/java/org/dspace/scripts/ProcessServiceImpl.java]: Yes, my bad. I changed it with **CommandLineDspaceRunnableHandler with command-line use..**

File: dspace-api/src/main/java/org/dspace/scripts/handler/impl/CommandLineDSpaceRunnableHandler.java [hunk 3]
```diff
@@ -10,34 +10,115 @@
...
+
 
     @Override
     public void start() {
         System.out.println("The script has started");
+        if (IS_SAVE_ENABLED) {
+            Context context = new Context();
+            try {
+                Process process = processService.find(context, processId);
+                processService.start(context, process);
```
[the remark this answers is not in the document]
Comment (Bezkup) [reply dspace-api/src/main/java/org/dspace/scripts/handler/impl/CommandLineDSpaceRunnableHandler.java:103]: You are right. I removed it because I was thinking that if the process is fast, the user will not see the status `RUNNING` but only `COMPLETED`. But if it takes a lot of time, the user might see only `NULL` and `COMPLETED` as `process.status`. Added the `context.complete()`

**Comment by tdonohue** [comment]
@Bezkup : Apologies for the delay in getting back... I hadn't had time to revisit this PR. 

I think you bring up a good point about the `./dspace database test` script (which is what Ant executes in a "fresh_install").  If I run that script manually, I also see it adds an entry to the Processes table in the Admin UI:

database-process

This is a **bit of a problem** because the `./dspace database` script CANNOT be run from the Admin UI.  It's not compatible with the Admin UI at this time. (Some are NOT meant to be run from the Admin UI.  This "database" script is too powerful & can run database migrations which is something you don't want to allow from the Admin UI... as it can potentially bring down the site.) 
 
Yet, if I click on that entry in the processes list, I see this:

database-process-details

There are two issues here:
1. I cannot create a similar process (button doesn't work as this cannot be run from the Admin UI)
2. It is refreshing over and over again as the Admin UI thinks this process is not complete.  It never stops refreshing.  If you look closely at that first screenshot you'll see it has an empty start/finish time & empty status. 

Overall, I think your simple question about this `./dspace database test` script is showing us that there's a flaw in the design here.  **I think we need to find a way to limit which scripts appear in the Processes page in the Admin UI.**   

More specifically, I'd suggest we may want to consider if there's a way to ONLY enable this feature for scripts which are visible/runnable in the Admin UI.  This likely means that you MUST remove the changes to `ScriptLauncher` in this PR, as scripts defined in `[dspace]/config/launcher.xml` are NOT runnable in the Admin UI.  Only scripts defined in `[dspace]/config/spring/api/scripts.xml` can be used *both* in the Admin UI and on CLI.   (In case it isn't clear, this would solve the "ant fresh_install" problem too as the `./dspace database` script is only in `launcher.xml`)

**Comment by Bezkup** [comment]
@tdonohue: Don't worry, we are all busy in this period :D 
Ok, I've tried to differentiate scripts within `scripts.xml` from the others within `launcher.xml`; I've used the class `ScriptService` to see if there are any `ScriptConfiguration` (as I understand, only scripts within scripts.xml have a `ScriptConfiguration` bean). [Here is](https://github.com/Bezkup/DSpace/blob/4ecbf74f0f2c69eb1132aa98caa6b4031621b283/dspace-api/src/main/java/org/dspace/scripts/handler/impl/CommandLineDSpaceRunnableHandler.java#L248) the function, tell me if you have a suggestion and how to improve it :)

**Comment by tdonohue** [comment]
@Bezkup : To clarify, I don't think you even need to add a function to determine which scripts are in `scripts.xml`.   While I could be wrong, I think the _only change_ that might be necessary is to remove all your modifications to `ScriptLauncher.java` in this PR

The reason is that we have two types of "scripts" currently in DSpace.
1. Scripts defined in `launcher.xml` which are always run via `ScriptLauncher.java`
2. Scripts defined in `scripts.xml` which are always run via `CommandLineDSpaceRunnableHandler.java` and the equivalent RunnableHandler in the REST API.

So, I think all we need to do is ensure this PR doesn't touch `ScriptLauncher.java`... that, in itself should ensure that scripts that appear in `launcher.xml` are never saved to the process table.  In other words, your new method shouldn't be necessary once your changes to ScriptLauncher are removed.

**Comment by Bezkup** [comment]
@tdonohue Mmm I don't know.. My idea with that function came up when I saw [this](https://github.com/Bezkup/DSpace/blob/19471b7a82300775dbeb442ec06dd9d473ab6eda/dspace-api/src/main/java/org/dspace/app/launcher/ScriptLauncher.java#L139). So `ScriptHandler` handles both scripts in `scripts.xml` and `launcher.xml`. Maybe I've not understand this step

**Comment by tdonohue** [comment]
@Bezkup : Sorry, I misspoke/mistyped slightly.   You are correct, that both types of scripts "flow through" ScriptLauncher.  But, the `ScriptLauncher.executeScript()` your referenced calls the `CommandLineDSpaceRunnableHandler` to run anything in the `scripts.xml` here: https://github.com/DSpace/DSpace/blob/main/dspace-api/src/main/java/org/dspace/app/launcher/ScriptLauncher.java#L149

What I'm trying to point out, is that, because the ScriptLauncher calls to  `CommandLineDSpaceRunnableHandler` for everything in `scripts.xml`, I don't think you need to actually touch the `ScriptLauncher`.  All you need to do is ensure that everything run via `CommandLineDSpaceRunnableHandler` is added to the processes table... that will already exclude all the processes in `launcher.xml` because they are run by the `ScriptLauncher.runOneCommand()` method.

**Comment by Bezkup** [comment]
@tdonohue Somehow I need to modify `ScriptLauncher` because if I don't pass the script name and arguments to `CommandLineDSpaceRunnableHandler` I cannot insert them in Process table. Can I retrieve them elsewhere?


> UPDATE: I should add, it's possible I'm overlooking something too. But I just find it odd that we need to check if something is in `scripts.xml` in **two places**. Once in `ScriptLauncher` (to decide whether to call `CommandLineDSpaceRunnableHandler`) and then a second time now in your new method. It seems like we should have already filtered out what is in `scripts.xml` with the first check that you referenced.
>
Yes, you're right. I was considering creating a `DspaceRunnable` and retrieve its configuration and pass it as a boolean (script != null) to the constructor of `CommandLineDSpaceRunnableHandler`. This would allow us to avoid to create the new function, and passing the `DspaceRunnable` (that might be null) to `ScriptLauncher.handleScript()` to determine which function to call (`executeScript()` or `runOneCommand()`). However, I believe it's not a good idea due to its complexity and the fact that this function is heavily used in many test classes.

Sorry if it sounds tricky but it's a bit complicated to manage :/

**Comment by Bezkup** [comment]
@tdonohue What if we move the logic of `Process` creation from `CommandLineDSpaceRunnableHandler` constructor to `CommandLineDSpaceRunnableHandler.start()` method? So we are 100% sure that `Process` will create only if the script is inside `scripts.xml` and avoid to call another function to verify this. 
But I still have to pass the script name and parameters to constructor so I can create the `Process` when `start()` method is called

**Comment by tdonohue** [comment]
@Bezkup : I fear I may have misled you...or gotten things confused.  I won't have immediate time this week to dig deep on this.  But, my main concern is simply that, prior to this PR, **I believe that only scripts in `scripts.xml` ran via `CommandLineDSpaceRunnableHandler`**.   This is the primary way that we "filtered out" scripts from `launcher.xml` appearing in the Admin UI.

So, assuming that is true, I believe that we already have a way to ensure that scripts in `launcher.xml` do NOT appear in the Admin UI.  All that you therefore need to do in this PR is ensure that scripts from `scripts.xml` run from CLI also _now appear_ in the Admin UI.   You are correct that this still may require some modifications to `ScriptLauncher`.  However, you need to find a way to ensure modifications to `ScriptLauncher` do NOT get run by scripts in `launcher.xml`...as you only need to deal with the scripts in `scripts.xml`. (E.g. you are modifying `ScriptLauncher.main()` which both types of scripts use... while I understand the reasons why, it's odd that scripts in `launcher.xml` are also running this code.  Shouldn't some/all of that code move into the "executeScript()" method which will ONLY be run by scripts in `scripts.xml`?)

I hope that makes some sense.  I'm swamped this week, so I don't know that I can provide more exact advice.  I'm just worried that some of the code in this PR is **unnecessarily** changing the behavior of scripts in `launcher.xml`, when all it needs to do is change the behavior of scripts in `scripts.xml`.   If you find a cleaner solution that you feel works well, I'm OK with it.

**Comment by Bezkup** [comment]
I believe this [commit](https://github.com/DSpace/DSpace/pull/8946/commits/d18271963421bda9b114a5fbc8d663b7655ede53) is the cleanest way to achieve this. I attempted to overload the `ScriptLauncher.handleScript` to accept only `args` e `commandArgs` and create a `CommandLineDspaceHandler` solely when the script is in `scripts.xml`. But the older function were only used in test classes, which makes it useless

Let me know if you have a suggestion, because I've run out of ideas :/

**Review by abollini — COMMENTED** [review]
Hi @Bezkup,
thanks for your willingness to contribute to the project. I'm in favor of this feature that in my opinion should be enabled by default. That said, I think that there are still few side effects that make this contribution not yet ready for 8.0. The current script & process architecture allows us to accept different parameters in the CLI respect to the ones allowed in the UI. This mean that we could list in the process table a script that has used parameters not allowed from the UI with the consequence that the "Create a similar process" button will still not work as previously reported by @tdonohue. I would prefer to have all the executed process listed in the processes table, improving the UI so that it is able to distinguish between process that can be re-run from the UI and process that can only be seen.
I would suggest you to take a look at the https://github.com/DSpace/DSpace/issues/9359 that also require fixing the eperson_id column in the process table to be nullable and provide a PR for just this fix. In this way we should be able to review and merge the database changes already in DSpace 8 so that this feature will become suitable for a 8.1 release.

**Comment by Bezkup** [comment]
@abollini  Hi! Alright then, I'll make another PR with only nullable. I'm happy to wait for any information you have about this PR :)

**Comment by kshepherd** [comment]
@Bezkup i re-ran the integration test pipeline and it has now passed :)

**Comment by Bezkup** [comment]
Closing this PR. I will reopen it when the time comes :)
